File: Intersection.java
// Time Complexity : O(M + N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :Y
// Any problem you faced while coding this :

public class Intersection {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int lenA = 0;
        int lenB = 0;
        ListNode curr = headA;
        while(curr !=null)
        {
            lenA++;
            curr = curr.next;
        }
        curr = headB;
        while(curr !=null)
        {
            lenB++;
            curr = curr.next;
        }
        while(lenA > lenB)
        {
            headA = headA.next;
            lenA--;
        }
        while(lenB > lenA)
        {
            headB = headB.next;
            lenB--;
        }
        while(headA != headB)
        {
            headA = headA.next;
            headB = headB.next;
        }
        return headB;
}

public static void main(String args [])
{
    Intersection list = new Intersection();
    ListNode node = new ListNode(4);
    node.next = new ListNode(1);
    node.next.next = new ListNode(8);
    node.next.next.next = new ListNode(4);
    node.next.next.next.next = new ListNode(5);
    ListNode second = new ListNode(5);
    second.next = new ListNode(6);
    second.next.next = new ListNode(1);
    second.next.next.next =  node.next.next;
    System.out.println(list.getIntersectionNode(node,second).val);
}

}

File: intersection.cpp
// Time Complexity : O(m+n);  m, n are numbers of nodes in each list respectively 
// Space Complexity : O(m)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// 1. Maintain set of nodes and add each node of first list 
// 2. Traverse through second list and see if any node is found in set already
// 3. If found return that node else return nullptr

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        set<ListNode*> mapper;
        ListNode* itr = headA;
        while(itr != nullptr){
            mapper.insert(itr);
            itr=itr->next;
        }
        itr = headB;
        while(itr!=nullptr){
            if(mapper.find(itr) != mapper.end())
                return itr;
            itr=itr->next;
        }
        return itr;
    }
};